File: frontend/src/lib/chemicalTreatment.ts
/**
 * Chemical Water Treatment — TypeScript types, label maps, API client.
 *
 * Data flows:
 *  water_treatment_chemicals   → chemical master catalog
 *  treatment_chemical_records  → dosing transaction / stock movement log
 */
import { apiClient } from "./api";

// ── Enums & label maps ────────────────────────────────────────────────────────

export type ChemicalCategory =
  | "BOILER_TREATMENT"
  | "OXYGEN_SCAVENGER"
  | "ANTISCALANT"
  | "CORROSION_INHIBITOR"
  | "BIOCIDE"
  | "COAGULANT"
  | "FLOCCULANT"
  | "PH_ADJUSTER"
  | "NEUTRALIZER"
  | "DEFOAMER"
  | "CHLORINE"
  | "DECHLORINATION"
  | "BIO_NUTRIENT"
  | "OTHER";

export const CHEMICAL_CATEGORY_LABELS: Record<ChemicalCategory, string> = {
  BOILER_TREATMENT:    "Boiler Treatment",
  OXYGEN_SCAVENGER:    "Oxygen Scavenger",
  ANTISCALANT:         "Antiscalant",
  CORROSION_INHIBITOR: "Corrosion Inhibitor",
  BIOCIDE:             "Biocide",
  COAGULANT:           "Coagulant",
  FLOCCULANT:          "Flocculant",
  PH_ADJUSTER:         "pH Adjuster",
  NEUTRALIZER:         "Neutralizer",
  DEFOAMER:            "Defoamer",
  CHLORINE:            "Chlorine / Hypochlorite",
  DECHLORINATION:      "Dechlorination",
  BIO_NUTRIENT:        "Biological Nutrient",
  OTHER:               "Other",
};

export const CHEMICAL_CATEGORY_COLORS: Record<ChemicalCategory, string> = {
  BOILER_TREATMENT:    "text-orange-400",
  OXYGEN_SCAVENGER:    "text-blue-400",
  ANTISCALANT:         "text-cyan-400",
  CORROSION_INHIBITOR: "text-amber-400",
  BIOCIDE:             "text-red-400",
  COAGULANT:           "text-violet-400",
  FLOCCULANT:          "text-purple-400",
  PH_ADJUSTER:         "text-emerald-400",
  NEUTRALIZER:         "text-teal-400",
  DEFOAMER:            "text-sky-400",
  CHLORINE:            "text-yellow-400",
  DECHLORINATION:      "text-lime-400",
  BIO_NUTRIENT:        "text-green-400",
  OTHER:               "text-slate-400",
};

export const CHEMICAL_CATEGORY_BG: Record<ChemicalCategory, string> = {
  BOILER_TREATMENT:    "bg-orange-500/10 border-orange-500/20",
  OXYGEN_SCAVENGER:    "bg-blue-500/10 border-blue-500/20",
  ANTISCALANT:         "bg-cyan-500/10 border-cyan-500/20",
  CORROSION_INHIBITOR: "bg-amber-500/10 border-amber-500/20",
  BIOCIDE:             "bg-red-500/10 border-red-500/20",
  COAGULANT:           "bg-violet-500/10 border-violet-500/20",
  FLOCCULANT:          "bg-purple-500/10 border-purple-500/20",
  PH_ADJUSTER:         "bg-emerald-500/10 border-emerald-500/20",
  NEUTRALIZER:         "bg-teal-500/10 border-teal-500/20",
  DEFOAMER:            "bg-sky-500/10 border-sky-500/20",
  CHLORINE:            "bg-yellow-500/10 border-yellow-500/20",
  DECHLORINATION:      "bg-lime-500/10 border-lime-500/20",
  BIO_NUTRIENT:        "bg-green-500/10 border-green-500/20",
  OTHER:               "bg-slate-500/10 border-slate-500/20",
};

export type TreatmentType =
  | "COAGULATION"
  | "FLOCCULATION"
  | "DISINFECTION"
  | "SOFTENING"
  | "REVERSE_OSMOSIS"
  | "BIOLOGICAL"
  | "PH_ADJUSTMENT"
  | "DECHLORINATION"
  | "FILTRATION"
  | "OTHER";

export const TREATMENT_TYPE_LABELS: Record<TreatmentType, string> = {
  COAGULATION:    "Coagulation",
  FLOCCULATION:   "Flocculation",
  DISINFECTION:   "Disinfection",
  SOFTENING:      "Softening",
  REVERSE_OSMOSIS: "Reverse Osmosis",
  BIOLOGICAL:     "Biological",
  PH_ADJUSTMENT:  "pH Adjustment",
  DECHLORINATION: "Dechlorination",
  FILTRATION:     "Filtration",
  OTHER:          "Other",
};

export type DosingMode = "MANUAL" | "AUTO";

// ── Chemical Master ────────────────────────────────────────────────────────────

export interface WaterTreatmentChemical {
  id:               string;
  chemical_code:    string;
  chemical_name:    string;
  chemical_category: ChemicalCategory;
  description?:     string | null;
  supplier_id?:     string | null;
  supplier_name?:   string | null;
  stock_uom:        string;
  dosing_uom:       string;
  unit_cost?:       number | null;
  target_dose_min?: number | null;
  target_dose_max?: number | null;
  low_stock_threshold?: number | null;
  material_id?:     string | null;
  is_active:        boolean;
  notes?:           string | null;
  created_at?:      string | null;
}

export interface WaterTreatmentChemicalCreate {
  chemical_code:    string;
  chemical_name:    string;
  chemical_category: ChemicalCategory;
  description?:     string;
  supplier_id?:     string;
  stock_uom:        string;
  dosing_uom:       string;
  unit_cost?:       number;
  target_dose_min?: number;
  target_dose_max?: number;
  low_stock_threshold?: number;
  material_id?:     string;
  is_active?:       boolean;
  notes?:           string;
}

export interface WaterTreatmentChemicalUpdate extends Partial<Omit<WaterTreatmentChemicalCreate, "chemical_code">> {}

export interface ChemicalOption {
  id:               string;
  chemical_code:    string;
  chemical_name:    string;
  chemical_category: ChemicalCategory;
  stock_uom:        string;
  dosing_uom:       string;
  unit_cost?:       number | null;
}

// ── Treatment Records ─────────────────────────────────────────────────────────

export interface TreatmentRecord {
  id:               string;
  record_no:        string;
  asset_id:         string;
  asset_name?:      string | null;
  asset_no?:        string | null;

  record_datetime:  string;
  date?:            string | null;
  treatment_type:   TreatmentType;

  chemical_id?:      string | null;
  chemical_code?:    string | null;
  chemical_name:     string;
  chemical_category?: ChemicalCategory | null;

  treatment_area?:   string | null;
  dosing_point?:     string | null;

  supplier_id?:      string | null;
  supplier_name?:    string | null;
  batch_lot_no?:     string | null;
  unit_cost?:        number | null;

  stock_uom?:        string | null;
  dosing_uom?:       string | null;

  opening_stock?:    number | null;
  received_qty?:     number | null;
  consumed_qty?:     number | null;
  closing_stock?:    number | null;

  quantity_dosed:    number;
  unit:              string;
  target_dose_ppm?:  number | null;
  actual_dose_ppm?:  number | null;
  water_volume_m3?:  number | null;

  manual_or_auto_dose?: DosingMode | null;
  related_meter_id?: string | null;

  water_treated_m3?:   number | null;
  steam_produced_ton?: number | null;

  feed_ph?:               number | null;
  product_ph?:            number | null;
  feed_turbidity_ntu?:    number | null;
  product_turbidity_ntu?: number | null;
  residual_chlorine_ppm?: number | null;
  tds_feed_ppm?:          number | null;
  tds_product_ppm?:       number | null;

  source_method:    string;
  is_anomaly:       boolean;
  anomaly_note?:    string | null;
  overdose_flag:    boolean;
  underdose_flag:   boolean;
  shift_ref?:       string | null;
  inventory_synced: boolean;
  notes?:           string | null;

  created_at?:      string | null;
  updated_at?:      string | null;
}

export interface TreatmentRecordCreate {
  asset_id:          string;
  record_datetime:   string;
  date?:             string;
  treatment_type:    TreatmentType;
  chemical_id?:      string;
  chemical_code?:    string;
  chemical_name:     string;
  chemical_category?: ChemicalCategory;
  treatment_area?:   string;
  dosing_point?:     string;
  supplier_id?:      string;
  batch_lot_no?:     string;
  unit_cost?:        number;
  stock_uom?:        string;
  dosing_uom?:       string;
  opening_stock?:    number;
  received_qty?:     number;
  consumed_qty?:     number;
  closing_stock?:    number;
  quantity_dosed:    number;
  unit:              string;
  target_dose_ppm?:  number;
  actual_dose_ppm?:  number;
  water_volume_m3?:  number;
  manual_or_auto_dose?: DosingMode;
  related_meter_id?: string;
  water_treated_m3?:   number;
  steam_produced_ton?: number;
  feed_ph?:               number;
  product_ph?:            number;
  feed_turbidity_ntu?:    number;
  product_turbidity_ntu?: number;
  residual_chlorine_ppm?: number;
  tds_feed_ppm?:          number;
  tds_product_ppm?:       number;
  source_method?:    string;
  is_anomaly?:       boolean;
  anomaly_note?:     string;
  overdose_flag?:    boolean;
  underdose_flag?:   boolean;
  shift_ref?:        string;
  notes?:            string;
}

export type TreatmentRecordUpdate = Partial<TreatmentRecordCreate>;

// ── KPIs ──────────────────────────────────────────────────────────────────────

export interface ChemicalKPIs {
  date_from?:              string | null;
  date_to?:                string | null;
  total_records:           number;
  total_quantity_dosed?:   number | null;
  total_consumed_qty?:     number | null;
  total_received_qty?:     number | null;
  total_cost?:             number | null;
  chemical_per_m3_water?:  number | null;
  chemical_per_ton_steam?: number | null;
  overdose_count:          number;
  underdose_count:         number;
  anomaly_count:           number;
  low_stock_chemicals:     string[];
}

// ── Filters ───────────────────────────────────────────────────────────────────

export interface ChemTreatmentFilters {
  date_from?:         string;
  date_to?:           string;
  asset_id?:          string;
  chemical_id?:       string;
  chemical_category?: ChemicalCategory;
  treatment_area?:    string;
  supplier_id?:       string;
  is_anomaly?:        boolean;
  overdose_flag?:     boolean;
  underdose_flag?:    boolean;
  shift_ref?:         string;
  batch_lot_no?:      string;
  skip?:              number;
  limit?:             number;
}

// ── API client ────────────────────────────────────────────────────────────────

export const chemTreatmentApi = {
  // KPIs
  kpis: (filters: Partial<ChemTreatmentFilters>) =>
    apiClient
      .get<ChemicalKPIs>("/chemical-treatment/kpis", { params: filters })
      .then((r) => r.data),

  // Chemical master
  chemicals: (params?: { chemical_category?: string; is_active?: boolean; search?: string }) =>
    apiClient
      .get<WaterTreatmentChemical[]>("/chemical-treatment/chemicals", { params })
      .then((r) => r.data),

  chemicalOptions: () =>
    apiClient
      .get<ChemicalOption[]>("/chemical-treatment/chemicals/options")
      .then((r) => r.data),

  getChemical: (id: string) =>
    apiClient
      .get<WaterTreatmentChemical>(`/chemical-treatment/chemicals/${id}`)
      .then((r) => r.data),

  createChemical: (data: WaterTreatmentChemicalCreate) =>
    apiClient
      .post<WaterTreatmentChemical>("/chemical-treatment/chemicals", data)
      .then((r) => r.data),

  updateChemical: (id: string, data: WaterTreatmentChemicalUpdate) =>
    apiClient
      .patch<WaterTreatmentChemical>(`/chemical-treatment/chemicals/${id}`, data)
      .then((r) => r.data),

  deleteChemical: (id: string) =>
    apiClient.delete(`/chemical-treatment/chemicals/${id}`),

  // Treatment records
  records: (filters: ChemTreatmentFilters) =>
    apiClient
      .get<TreatmentRecord[]>("/chemical-treatment/records", { params: filters })
      .then((r) => r.data),

  getRecord: (id: string) =>
    apiClient
      .get<TreatmentRecord>(`/chemical-treatment/records/${id}`)
      .then((r) => r.data),

  createRecord: (data: TreatmentRecordCreate) =>
    apiClient
      .post<TreatmentRecord>("/chemical-treatment/records", data)
      .then((r) => r.data),

  updateRecord: (id: string, data: TreatmentRecordUpdate) =>
    apiClient
      .patch<TreatmentRecord>(`/chemical-treatment/records/${id}`, data)
      .then((r) => r.data),

  deleteRecord: (id: string) =>
    apiClient.delete(`/chemical-treatment/records/${id}`),

  // Assets dropdown
  assets: () =>
    apiClient
      .get<{ id: string; asset_no: string; name: string; utility_type: string }[]>(
        "/chemical-treatment/assets"
      )
      .then((r) => r.data),
};

export const downloadChemicalTreatmentCsv = async (filters: ChemTreatmentFilters) => {
  const response = await apiClient.get("/chemical-treatment/records/export/csv", {
    params: filters,
    responseType: "blob",
  });
  const url = window.URL.createObjectURL(response.data);
  const a = document.createElement("a");
  a.href = url;
  a.download = `chemical-treatment-${new Date().toISOString().slice(0, 10)}.csv`;
  a.click();
  window.URL.revokeObjectURL(url);
};

File: backend/app/schemas/chemical_treatment.py
"""
Pydantic schemas for Chemical Water Treatment module.
"""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, field_validator, model_validator


# ── Chemical Master ────────────────────────────────────────────────────────────

class WaterTreatmentChemicalBase(BaseModel):
    chemical_code:    str
    chemical_name:    str
    chemical_category: str
    description:      Optional[str]  = None
    supplier_id:      Optional[UUID] = None
    stock_uom:        str
    dosing_uom:       str
    unit_cost:        Optional[Decimal] = None
    target_dose_min:  Optional[Decimal] = None
    target_dose_max:  Optional[Decimal] = None
    low_stock_threshold: Optional[Decimal] = None
    material_id:      Optional[UUID] = None
    is_active:        bool = True
    notes:            Optional[str]  = None

    @field_validator("chemical_category")
    @classmethod
    def _chk_category(cls, v: str) -> str:
        valid = {
            "BOILER_TREATMENT", "OXYGEN_SCAVENGER", "ANTISCALANT",
            "CORROSION_INHIBITOR", "BIOCIDE", "COAGULANT", "FLOCCULANT",
            "PH_ADJUSTER", "NEUTRALIZER", "DEFOAMER", "CHLORINE",
            "DECHLORINATION", "BIO_NUTRIENT", "OTHER",
        }
        if v.upper() not in valid:
            raise ValueError(f"chemical_category must be one of {sorted(valid)}")
        return v.upper()


class WaterTreatmentChemicalCreate(WaterTreatmentChemicalBase):
    pass


class WaterTreatmentChemicalUpdate(BaseModel):
    chemical_name:    Optional[str]     = None
    chemical_category: Optional[str]    = None
    description:      Optional[str]     = None
    supplier_id:      Optional[UUID]    = None
    stock_uom:        Optional[str]     = None
    dosing_uom:       Optional[str]     = None
    unit_cost:        Optional[Decimal] = None
    target_dose_min:  Optional[Decimal] = None
    target_dose_max:  Optional[Decimal] = None
    low_stock_threshold: Optional[Decimal] = None
    material_id:      Optional[UUID]    = None
    is_active:        Optional[bool]    = None
    notes:            Optional[str]     = None


class WaterTreatmentChemicalRead(WaterTreatmentChemicalBase):
    id:            UUID
    supplier_name: Optional[str] = None
    created_at:    Optional[datetime] = None

    model_config = {"from_attributes": True}


# ── Treatment Records ──────────────────────────────────────────────────────────

class TreatmentRecordCreate(BaseModel):
    asset_id:          UUID
    record_datetime:   datetime
    date:              Optional[date]  = None
    treatment_type:    str

    chemical_id:       Optional[UUID]  = None
    chemical_code:     Optional[str]   = None
    chemical_name:     str
    chemical_category: Optional[str]   = None

    treatment_area:    Optional[str]   = None
    dosing_point:      Optional[str]   = None

    supplier_id:       Optional[UUID]  = None
    batch_lot_no:      Optional[str]   = None
    unit_cost:         Optional[Decimal] = None

    stock_uom:         Optional[str]   = None
    dosing_uom:        Optional[str]   = None

    opening_stock:     Optional[Decimal] = None
    received_qty:      Optional[Decimal] = Decimal("0")
    consumed_qty:      Optional[Decimal] = Decimal("0")
    closing_stock:     Optional[Decimal] = None

    quantity_dosed:    Decimal
    unit:              str
    target_dose_ppm:   Optional[Decimal] = None
    actual_dose_ppm:   Optional[Decimal] = None
    water_volume_m3:   Optional[Decimal] = None

    manual_or_auto_dose: Optional[str] = "MANUAL"

    related_meter_id:  Optional[UUID]  = None
    material_id:       Optional[UUID]  = None

    water_treated_m3:   Optional[Decimal] = None
    steam_produced_ton: Optional[Decimal] = None

    feed_ph:               Optional[Decimal] = None
    product_ph:            Optional[Decimal] = None
    feed_turbidity_ntu:    Optional[Decimal] = None
    product_turbidity_ntu: Optional[Decimal] = None
    residual_chlorine_ppm: Optional[Decimal] = None
    tds_feed_ppm:          Optional[Decimal] = None
    tds_product_ppm:       Optional[Decimal] = None

    source_method: str = "MANUAL"
    is_anomaly:    bool = False
    anomaly_note:  Optional[str] = None
    overdose_flag: bool = False
    underdose_flag: bool = False
    shift_ref:     Optional[str] = None
    notes:         Optional[str] = None

    @field_validator("treatment_type")
    @classmethod
    def _chk_treatment_type(cls, v: str) -> str:
        valid = {
            "COAGULATION", "FLOCCULATION", "DISINFECTION", "SOFTENING",
            "REVERSE_OSMOSIS", "BIOLOGICAL", "PH_ADJUSTMENT",
            "DECHLORINATION", "FILTRATION", "OTHER",
        }
        if v.upper() not in valid:
            raise ValueError(f"treatment_type must be one of {sorted(valid)}")
        return v.upper()

    @field_validator("manual_or_auto_dose")
    @classmethod
    def _chk_dosing_mode(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        if v.upper() not in {"MANUAL", "AUTO"}:
            raise ValueError("manual_or_auto_dose must be MANUAL or AUTO")
        return v.upper()

    @model_validator(mode="after")
    def _auto_fill_date(self) -> "TreatmentRecordCreate":
        if self.date is None and self.record_datetime:
            self.date = self.record_datetime.date()
        return self

    @model_validator(mode="after")
    def _detect_dose_anomaly(self) -> "TreatmentRecordCreate":
        """Auto-set overdose/underdose flags based on target_dose_ppm vs actual_dose_ppm."""
        if self.target_dose_ppm and self.actual_dose_ppm:
            tgt = float(self.target_dose_ppm)
            act = float(self.actual_dose_ppm)
            if act > tgt * 1.20:
                self.overdose_flag = True
            elif act < tgt * 0.80:
                self.underdose_flag = True
        return self


class TreatmentRecordUpdate(BaseModel):
    record_datetime:   Optional[datetime] = None
    date:              Optional[date]     = None
    treatment_type:    Optional[str]      = None

    chemical_id:       Optional[UUID]    = None
    chemical_code:     Optional[str]     = None
    chemical_name:     Optional[str]     = None
    chemical_category: Optional[str]     = None

    treatment_area:    Optional[str]     = None
    dosing_point:      Optional[str]     = None

    supplier_id:       Optional[UUID]    = None
    batch_lot_no:      Optional[str]     = None
    unit_cost:         Optional[Decimal] = None

    stock_uom:         Optional[str]     = None
    dosing_uom:        Optional[str]     = None

    opening_stock:     Optional[Decimal] = None
    received_qty:      Optional[Decimal] = None
    consumed_qty:      Optional[Decimal] = None
    closing_stock:     Optional[Decimal] = None

    quantity_dosed:    Optional[Decimal] = None
    unit:              Optional[str]     = None
    target_dose_ppm:   Optional[Decimal] = None
    actual_dose_ppm:   Optional[Decimal] = None
    water_volume_m3:   Optional[Decimal] = None

    manual_or_auto_dose: Optional[str]  = None

    related_meter_id:  Optional[UUID]   = None

    water_treated_m3:   Optional[Decimal] = None
    steam_produced_ton: Optional[Decimal] = None

    feed_ph:               Optional[Decimal] = None
    product_ph:            Optional[Decimal] = None
    feed_turbidity_ntu:    Optional[Decimal] = None
    product_turbidity_ntu: Optional[Decimal] = None
    residual_chlorine_ppm: Optional[Decimal] = None
    tds_feed_ppm:          Optional[Decimal] = None
    tds_product_ppm:       Optional[Decimal] = None

    source_method: Optional[str]  = None
    is_anomaly:    Optional[bool] = None
    anomaly_note:  Optional[str]  = None
    overdose_flag: Optional[bool] = None
    underdose_flag: Optional[bool] = None
    shift_ref:     Optional[str]  = None
    notes:         Optional[str]  = None


class TreatmentRecordRead(BaseModel):
    id:              UUID
    record_no:       str
    asset_id:        UUID
    asset_name:      Optional[str] = None
    asset_no:        Optional[str] = None

    record_datetime: datetime
    date:            Optional[date]   = None
    treatment_type:  str

    chemical_id:      Optional[UUID] = None
    chemical_code:    Optional[str]  = None
    chemical_name:    str
    chemical_category: Optional[str] = None

    treatment_area:   Optional[str] = None
    dosing_point:     Optional[str] = None

    supplier_id:      Optional[UUID] = None
    supplier_name:    Optional[str]  = None
    batch_lot_no:     Optional[str]  = None
    unit_cost:        Optional[Decimal] = None

    stock_uom:        Optional[str] = None
    dosing_uom:       Optional[str] = None

    opening_stock:    Optional[Decimal] = None
    received_qty:     Optional[Decimal] = None
    consumed_qty:     Optional[Decimal] = None
    closing_stock:    Optional[Decimal] = None

    quantity_dosed:   Decimal
    unit:             str
    target_dose_ppm:  Optional[Decimal] = None
    actual_dose_ppm:  Optional[Decimal] = None
    water_volume_m3:  Optional[Decimal] = None

    manual_or_auto_dose: Optional[str] = None

    related_meter_id: Optional[UUID] = None

    water_treated_m3:   Optional[Decimal] = None
    steam_produced_ton: Optional[Decimal] = None

    feed_ph:               Optional[Decimal] = None
    product_ph:            Optional[Decimal] = None
    feed_turbidity_ntu:    Optional[Decimal] = None
    product_turbidity_ntu: Optional[Decimal] = None
    residual_chlorine_ppm: Optional[Decimal] = None
    tds_feed_ppm:          Optional[Decimal] = None
    tds_product_ppm:       Optional[Decimal] = None

    source_method:  str
    is_anomaly:     bool
    anomaly_note:   Optional[str] = None
    overdose_flag:  bool
    underdose_flag: bool
    shift_ref:      Optional[str] = None
    inventory_synced: bool = False
    notes:          Optional[str] = None

    created_at:     Optional[datetime] = None
    updated_at:     Optional[datetime] = None

    model_config = {"from_attributes": True}


# ── KPIs ──────────────────────────────────────────────────────────────────────

class ChemicalKPIs(BaseModel):
    date_from:                Optional[date]    = None
    date_to:                  Optional[date]    = None
    total_records:            int               = 0
    total_quantity_dosed:     Optional[Decimal] = None
    total_consumed_qty:       Optional[Decimal] = None
    total_received_qty:       Optional[Decimal] = None
    total_cost:               Optional[Decimal] = None
    chemical_per_m3_water:    Optional[Decimal] = None   # dosing unit per m3 treated water
    chemical_per_ton_steam:   Optional[Decimal] = None   # dosing unit per ton steam
    overdose_count:           int               = 0
    underdose_count:          int               = 0
    anomaly_count:            int               = 0
    low_stock_chemicals:      List[str]         = []


# ── Trend point ───────────────────────────────────────────────────────────────

class ChemicalTrendPoint(BaseModel):
    day:             str
    quantity_dosed:  Optional[Decimal] = None
    consumed_qty:    Optional[Decimal] = None
    overdose_count:  int = 0
    underdose_count: int = 0


# ── Dropdown ─────────────────────────────────────────────────────────────────

class ChemicalOption(BaseModel):
    id:               UUID
    chemical_code:    str
    chemical_name:    str
    chemical_category: str
    stock_uom:        str
    dosing_uom:       str
    unit_cost:        Optional[Decimal] = None
